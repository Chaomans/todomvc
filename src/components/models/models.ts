export type Filter = {
    name: string,
    value: string
}

export type TodoItem = {
    id: string;
    description: string;
    done: boolean;
}