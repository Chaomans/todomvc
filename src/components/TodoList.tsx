import React, { useState } from "react";
import Todo from "./Todo";
import TodoFooter from "./TodoFooter";
import TodoInput from "./TodoInput";

interface Itodo {
    id: number;
    description: string;
    done: boolean;
}

const TodoList = () => {
    const [todoList, setTodoList] = useState<Itodo[]>([
        { id: 0, description: "finish this todo", done: true },
        { id: 1, description: "finish this todo too", done: false },
    ]);

    const [newValue, setNewValue] = useState<string>("");

    const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        setNewValue(e.target.value);
    };

    const createTodo = (event: React.KeyboardEvent<HTMLInputElement>) => {
        if (event.key === "Enter" && newValue.length >= 3) {
            setTodoList([
                ...todoList,
                { id: Date.now(), description: newValue, done: false },
            ]);
            const inputEL = document.querySelector(
                ".new-todo"
            ) as HTMLInputElement;
            inputEL.value = "";
            setNewValue("");
        }
    };

    const deleteTodo = (id: number) => {
        setTodoList(todoList.filter((todo) => todo.id !== id));
    };

    const onChangeDone = (id: number) => {
        const doneTodo = todoList.find((todo) => todo.id === id) ?? null;
        if (doneTodo === null) return;
        const doneTodoIndex =
            todoList.findIndex((todo) => todo.id === id) ?? null;
        const newTodoList: Itodo[] = [];
        todoList.forEach((todo, i) => {
            if (i === doneTodoIndex) {
                newTodoList.push({
                    ...todoList[doneTodoIndex],
                    done: !doneTodo.done,
                });
            } else {
                newTodoList.push(todo);
            }
        });
        setTodoList(newTodoList);
    };

    const onEditDescription = (id: number, desc: string) => {
        const toEditIndex =
            todoList.findIndex((todo) => todo.id === id) ?? null;
        if (toEditIndex === null) return;
        const newTodoList: Itodo[] = [];
        todoList.forEach((todo, i) => {
            if (i === toEditIndex) {
                newTodoList.push({
                    ...todoList[toEditIndex],
                    description: desc,
                });
            } else {
                newTodoList.push(todo);
            }
        });
        setTodoList(newTodoList);
    };

    const onToggleAll = () => {
        const alldone = todoList.every((todo) => todo.done);
        const newTodoList: Itodo[] = [];
        if (alldone) {
            todoList.forEach((todo) =>
                newTodoList.push({
                    ...todo,
                    done: false,
                })
            );
        } else {
            todoList.forEach((todo) =>
                newTodoList.push({
                    ...todo,
                    done: true,
                })
            );
        }
        setTodoList(newTodoList);
    };

    const onClearCompleted = () => {
        setTodoList(todoList.filter((todo) => !todo.done));
    };

    return (
        <section className="todoapp">
            <TodoInput createTodo={createTodo} onChange={onChange} />
            <section className="main">
                <input
                    type="checkbox"
                    id="toggle-all"
                    className="toggle-all"
                    onChange={() => onToggleAll()}
                />
                <label htmlFor="toggle-all">Mark all as completed</label>
                <ul className="todo-list">
                    {todoList.map((todo: Itodo) => (
                        <Todo
                            key={todo.id}
                            id={todo.id}
                            description={todo.description}
                            onEditDescription={onEditDescription}
                            onDelete={deleteTodo}
                            onChangeDone={onChangeDone}
                            isdone={todo.done}
                        />
                    ))}
                </ul>
            </section>
            {todoList.length > 0 && (
                <TodoFooter
                    nbTodo={todoList.length}
                    onClearCompleted={onClearCompleted}
                />
            )}
        </section>
    );
};

export default TodoList;
