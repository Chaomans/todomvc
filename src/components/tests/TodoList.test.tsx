import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { TodoItem } from "../models/models";
import TodoList from "../TodoList";

describe("TodoList component", () => {

    it("Should render properly", () => {
        const todoList: TodoItem[] = [
            {id: "1A2Z", description: "Eat a rainbow", done: false},
            {id: "2Z3E", description: "Play rhythm game", done: true},
        ]

        const { asFragment } = render(
            <TodoList
                todos={todoList}
                onEditDescription={() => {}}
                onChangeDoneValue={() => {}}
                onDelete={() => {}}
            />
        )

        expect(asFragment()).toMatchSnapshot();
    })

    it("Should give 'editing' class to li when enter in editing mode", async () => {
        const todo: TodoItem = {
            id: "1A2Z",
            description: "Eat a rainbow",
            done: false
        };

        render(
            <TodoList
            todos={[todo]}
            onEditDescription={() => {}}
            onChangeDoneValue={() => {}}
            onDelete={() => {}}
            />
        )

        await userEvent.dblClick(await screen.findByText(todo.description));
        expect(await screen.findByRole("listitem")).toHaveClass("editing");
    })

    it("Should notify parent to update the description if the edition is valid", async () => {
        const mockOnEditDescription = jest.fn();

        const todo: TodoItem = {
            id: "1A2Z",
            description: "Eat a rainbow",
            done: false
        };

        render(
            <TodoList
            todos={[todo]}
            onEditDescription={mockOnEditDescription}
            onChangeDoneValue={() => {}}
            onDelete={() => {}}
            />
        )
        
        await userEvent.dblClick(await screen.findByText(todo.description));
        await userEvent.type(
            await screen.findByRole("textbox"),
            `${'{backspace}'.repeat(7)}burger{enter}`
        );
        expect(mockOnEditDescription).toHaveBeenCalledTimes(1);
        expect(mockOnEditDescription).toHaveBeenCalledWith(todo.id, "Eat a burger");
    });

    it("Should notify parent to delete the todo if the edition is invalid", async () => {
        const mockOnEditDescription = jest.fn();
        const mockOnDelete = jest.fn();

        const todo: TodoItem = {
            id: "1A2Z",
            description: "Eat a rainbow",
            done: false
        };

        render(
            <TodoList
            todos={[todo]}
            onEditDescription={mockOnEditDescription}
            onChangeDoneValue={() => {}}
            onDelete={mockOnDelete}
            />
        )
        
        await userEvent.dblClick( await screen.findByText(todo.description));
        await userEvent.type(
            await screen.findByRole("textbox"),
            `${'{backspace}'.repeat(todo.description.length)}   \t\t{Enter}`
        );
        expect(mockOnEditDescription).not.toHaveBeenCalled();
        expect(mockOnDelete).toHaveBeenCalledTimes(1);
        expect(mockOnDelete).toHaveBeenCalledWith(todo.id);
    });

    it("Should remove editing class if escape is pressed", async () => {
        const mockOnEditDescription = jest.fn();
        const todo: TodoItem = {
            id: "1A2Z",
            description: "Eat a rainbow",
            done: false
        };

        render(
            <TodoList
            todos={[todo]}
            onEditDescription={mockOnEditDescription}
            onChangeDoneValue={() => {}}
            onDelete={() => {}}
            />
        )

        
        await userEvent.dblClick( await screen.findByText(todo.description));
        await userEvent.type(await screen.findByRole("textbox"), "insert smth");
        await userEvent.keyboard('{Escape}');
        expect(mockOnEditDescription).not.toHaveBeenCalled();
        expect(await screen.findByRole("listitem")).not.toHaveClass("editing");
    });
    
    it("Should notify parent on delete", async () => {
        const mockOnDelete = jest.fn();
        const todo: TodoItem = {
            id: "1A2Z",
            description: "Eat a rainbow",
            done: false
        };

        render(
            <TodoList
            todos={[todo]}
            onEditDescription={() => {}}
            onChangeDoneValue={() => {}}
            onDelete={mockOnDelete}
            />
        )

        await userEvent.click(await screen.findByRole("button"));
        expect(mockOnDelete).toHaveBeenCalledTimes(1);
        expect(mockOnDelete).toHaveBeenCalledWith(todo.id);
    })

    it.each`
    completed  | status
    ${true}    | ${"completed"}
    ${false}    | ${"active"}
    `("Should prevent parent to change status from $status", async ({completed}) => {
        const mockOnChangeDoneValue = jest.fn();
        const todo: TodoItem = {
            id: "1A2Z",
            description: "Eat a rainbow",
            done: completed
        };

        render(
            <TodoList
            todos={[todo]}
            onEditDescription={() => {}}
            onChangeDoneValue={mockOnChangeDoneValue}
            onDelete={() => {}}
            />
        )

        await userEvent.click(await screen.findByRole("checkbox"));
        expect(mockOnChangeDoneValue).toHaveBeenCalledTimes(1);
        expect(mockOnChangeDoneValue).toHaveBeenCalledWith(todo.id, !completed);
    })

});