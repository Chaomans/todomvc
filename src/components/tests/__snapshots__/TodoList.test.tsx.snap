// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList component Should render properly 1`] = `
<DocumentFragment>
  <ul
    class="todo-list"
  >
    <li
      class=""
    >
      <div
        class="view"
      >
        <input
          class="toggle"
          type="checkbox"
        />
        <label>
          Eat a rainbow
        </label>
        <button
          class="destroy"
        />
      </div>
    </li>
    <li
      class="completed"
    >
      <div
        class="view"
      >
        <input
          checked=""
          class="toggle"
          type="checkbox"
        />
        <label>
          Play rhythm game
        </label>
        <button
          class="destroy"
        />
      </div>
    </li>
  </ul>
</DocumentFragment>
`;
